src/ddspdrum/torchmodule.py

```diff
@@ -486,12 +529,164 @@ def _forward(self, control_in: T, audio_in: T) -> T:
+
+    @staticmethod
+    def noise_of_length(audio_in: T) -> T:
+        return torch.rand_like(audio_in) * 2 - 1
+
+
+class TorchSynth(nn.Module):
+    """
+    Base class for synthesizers that combine one or more TorchSynthModules
+    to create a full synth architecture.
+
+    Parameters
+    ----------
+    sample_rate (int): sample rate to run this synth at
+    buffer_size (int): number of samples expected at output of child modules
+    """
+
+    def __init__(
+            self,
+            sample_rate: int = SAMPLE_RATE,
+            buffer_size: int = BUFFER_SIZE
+    ):
+        super().__init__()
+        self.sample_rate = T(sample_rate)
+        self.buffer_size = T(buffer_size)
+
+    def add_synth_modules(self, modules: Dict[str, TorchSynthModule]):
+        """
+        Add a set of named children TorchSynthModules to this synth. Registers them
+        with the torch nn.Module so that all parameters are recognized.
+
+        Parameters
+        ----------
+        modules (Dict): A dictionary of TorchSynthModule
+        """
+
+        for name in modules:
+            if not isinstance(modules[name], TorchSynthModule):
+                raise TypeError(f"{modules[name]} is not a TorchSynthModule")
+
+            if modules[name].sample_rate != self.sample_rate:
+                raise ValueError(f"{modules[name]} sample rate does not match")
+
+            if modules[name].buffer_size != self.buffer_size:
+                raise ValueError(f"{modules[name]} buffer size does not match")
+
+            self.add_module(name, modules[name])
+
+    def randomize(self):
+        """
+        Randomize all parameters
+        """
+        for parameter in self.parameters():
+            parameter.data = torch.rand_like(parameter)
+
+
+class TorchDrum(TorchSynth):
+    """
+    A package of modules that makes one drum hit.
+    """
+
+    def __init__(
+        self,
+        note_on_duration: float,
+        vco_ratio: float = 0.5,
+        pitch_adsr: TorchADSR = TorchADSR(),
+        amp_adsr: TorchADSR = TorchADSR(),
+        vco_1: TorchVCO = TorchSineVCO(),
+        vco_2: TorchVCO = TorchSquareSawVCO(),
+        noise: TorchNoise = TorchNoise(),
+        vca: TorchVCA = TorchVCA(),
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        assert note_on_duration >= 0
+
+        # We assume that sustain duration is a hyper-parameter,
+        # with the mindset that if you are trying to learn to
+        # synthesize a sound, you won't be adjusting the note_on_duration.
+        # However, this is easily changed if desired.
+        self.note_on_duration = T(note_on_duration)
+
+        # Setup a dummy global parameter module
+        global_params = TorchSynthModule(
+            self.sample_rate.item(),
+            self.buffer_size.item()
+        )
+        global_params.add_parameters([
+            TorchParameter(vco_ratio, "vco_ratio", ParameterRange())
+        ])
```

Comment:
Cool. Can't help but wonder about other design options here. This works for sure. Just feels a little, I dunno, my spidey sense is tingling. If you feel likewise, great — if not, happy to leave it a move on.

Reply:
@maxsolomonhenry can you explain the design question here?